bessctl/measurement_utils.py

```diff
@@ -0,0 +1,393 @@
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import print_function
+
+import copy
+import os
+import sys
+import time
+
+try:
+    this_dir = os.path.dirname(os.path.realpath(__file__))
+    bessctl = os.path.join(this_dir, 'bessctl')
+    sys.path.insert(1, os.path.join(this_dir, '../../../'))
+    from pybess.bess import *
+except ImportError:
+    print('Cannot import the API module (pybess)', file=sys.stderr)
+    raise
+
+
+def get_local_bess_handle():
+    bess = BESS()
+    try:
+        bess.connect()
+    except BESS.APIError:
+        raise Exception('BESS is not running')
+    return bess
+
+
+class PortStats(object):
+
+    """
+    Stores PMDPort stats and optional Measure module output and exports some
+    acceessors for convenience.
+    """
+
+    def __init__(self, throughput, rtt, jitter):
+        self.throughput = throughput
+        self.rtt = rtt
+        self.jitter = jitter
+
+    def pkts_in(self):
+        return self.throughput.inc.packets
+
+    def pkts_out(self):
+        return self.throughput.out.packets
+
+    def drops_in(self):
+        return self.throughput.inc.dropped
+
+    def drops_out(self):
+        return self.throughput.out.dropped
+
+    def bytes_in(self):
+        return self.throughput.inc.bytes
+
+    def bytes_out(self):
+        return self.throughput.out.bytes
+
+    def rtt(self, percentile=None):
+        """
+        Returns the dictionary of all RTT value if `percentile` is none.
+        Otherwise returns the `percentile`th percentile RTT.
+        """
+        if self.rtt is None:
+            return None
+        if percentile is None:
+            return self.rtt
+        return self.rtt[percentile]
+
+    def jitter(self, percentile=None):
+        """
+        Returns the dictionary of all jitter value if `percentile` is none.
+        Otherwise returns the `percentile`th percentile jitter.
+        """
+        if jitter is None:
+            return None
+        if percentile is None:
+            return self.jitter
+        return self.jitter[percentile]
+
+    def __str__(self):
+        fmt = '[in / out] pkts: {:.3f}M / {:.3f}M, drops: {:.3f}M / {:.3f}M, bytes: {} / {}'
+        rtt_data = []
+        if self.rtt is not None:
+            fmt += '\n'
+            fmt += ', '.join(['rtt_{}: {{:.3f}} us'.format(k)
+                             for k in sorted(self.rtt.keys())])
+            for k in sorted(self.rtt.keys()):
```

Comment:
I hate to see `sorted(arg)` twice like this, and I guess we are guaranteed that if self.rtt is not None, self.jitter is also not None, but it might be nice to have a static method that takes a string-and-dict pair and computes the format and data from the sorted keys, or returns '' if they are None.  Then you could do:

    rtt_data = []
    fmt += self.format_data('rtt', self.rtt, rtt_data)
    fmt += self.format_data('jit', self.jitter, rtt_data)
    return fmt.format(fmt, ...)
and have the `@staticmethod` do a single sorted() and data-append.

Kind of sixes though.

Reply:
Fixed! :)